pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +518,152 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
+			numNotVisited++
+			break
+		}
+		derr := d.listVolumesByResourceGroup(ctx, resourceGroup, &entries, start, maxEntries, &numVisited, &numNotVisited, &numEntries, &volSet)
+		if derr != nil {
+			if status.Code(derr) == codes.FailedPrecondition {
+				continue
+			}
+			return nil, derr
+		}
+	}
+
+	nextTokenString := ""
+	if numNotVisited > 0 {
+		nextTokenString = strconv.Itoa(numVisited)
 	}
-	pageEnd := start + maxEntries
 
-	vmSet := d.cloud.VMSet
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesWithoutKubeClient is a helper function for ListVolumes used for when there is no available kubeclient
+func (d *Driver) listVolumesWithoutKubeClient(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
 	entries := []*csi.ListVolumesResponse_Entry{}
-	for i := start; i < pageEnd; i++ {
-		d := disks[i]
-		nodeList := []string{}
+	numVisited, numNotVisited, numEntries := 0, 0, 0
+	derr := d.listVolumesByResourceGroup(ctx, d.cloud.ResourceGroup, &entries, start, maxEntries, &numVisited, &numNotVisited, &numEntries, nil)
+	if derr != nil {
+		return nil, derr
+	}
+
+	nextTokenString := ""
+	if numNotVisited > 0 {
+		nextTokenString = strconv.Itoa(numVisited)
+	}
+
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesByResourceGroup is a helper function that updates the ListVolumeResponse_Entry slice and returns error if found
+func (d *Driver) listVolumesByResourceGroup(ctx context.Context, resourceGroup string, entries *[]*csi.ListVolumesResponse_Entry, start, maxEntries int, numVisited, numNotVisited, numEntries *int, volSet *map[string]bool) error {
+	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, resourceGroup)
+	if derr != nil {
+		return status.Errorf(codes.Internal, "ListVolumes on rg(%s) failed with error: %v", resourceGroup, derr.Error())
+	}
 
+	if start > *numVisited+len(disks) {
+		return status.Errorf(codes.FailedPrecondition, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
+	}
+	i := start - *numVisited
+	end := len(disks)
+	if maxEntries > 0 && maxEntries < len(disks)-i {
+		end = maxEntries
+	}
+	*numNotVisited = len(disks)
+	for ; i < end; i++ {
+		disk := disks[i]
+		// if given a set of volumes from KubeClient, only continue if the disk can be found in the set
+		if volSet != nil && (*volSet)[*disk.ID] {
+			continue
+		}
 		// HyperVGeneration property is only setup for os disks. Only the non os disks should be included in the list
-		if d.DiskProperties == nil || d.DiskProperties.HyperVGeneration == "" {
-			if d.ManagedBy != nil {
-				attachedNode, err := vmSet.GetNodeNameByProviderID(*d.ManagedBy)
+		if disk.DiskProperties == nil || disk.DiskProperties.HyperVGeneration == "" {
```

Comment:
If some of these disks are os disks, they will not be added to the list of entries. So the actual list of entries will be less than the maxEntries even if the len(disks) >= maxEntries. We should keep running the loop as long as there are disks to iterate over and len(entries) < maxEntries.